### PR TITLE
HPC-GAP: flush method cache if the list of methods changes

### LINKED ISSUE #2594: SET_METHODS_OPERATION does not flush the methods cache.
### Observed behaviour

tests in #2521 fail in HPCGAP and I can't see how to rewrite the critical function to make them pass

### Expected behaviour

There should be a way to flush the methods cache (of an operation and number of arguments)
when needed.

### Copy and paste GAP banner (to tell us about your setup)
```
 *********   HPC-GAP 4.10dev-862-g3c8b617 of today
 *  GAP  *   https://www.gap-system.org
 *********   Architecture: x86_64-apple-darwin17.6.0-hpcgap64
             Maximum concurrent threads: 4
 Configuration:  gmp 6.0.0

#W <<< This is an alpha release.      >>>
#W <<< Do not use for important work. >>>
```

It seems (looking at INSTALL_METHOD_FLAGS in oper1.g for instance, that `SET_METHODS_OPERATION(oper,n,MakeReadOnlySingleObj(meths));`                  
(after copying the old methods list into `meths` and editing it, should be the correct way to modify a methods list in HPCGAP, replacing editing it in-place and calling `CHANGED_METHODS_OPERATION(oper,n);` but this doesn't work. Nothing in `SET_METHODS_OPERATION` flushes the method cache.

### PR DESCRIPTION
**Pull request by fingolfin**
Fixes #2594

### CODE REVIEW

**Comment by fingolfin** [comment]
I was confused by a test failure in `tst/testinstall/memoize.tst`; that file states "HPCGAP currently disables flushing caches" (I think @ChrisJefferson added that).

Well, turns out HPC-GAP does not so much "disable" `FlushCaches`, but rather just breaks it, due to method caching... So this PR fixes `FlushCaches` in HPC-GAP as a side effect. Nice.

**Comment by ChrisJefferson** [comment]
I seem to remember we didn't want to enable flushcaches in hpcgap due to worries about how it would interact with threads. I think some cached are thread local, have you looked to see what flushcaches in hpcgap is now doing? I guess flushing the thread?

Sorry, going from memory and away from computer.

**Comment by fingolfin** [comment]
@ChrisJefferson it is possible that somebody wanted to disable flushing of caches (perhaps @stevelinton or @rbehrends know or remember). But nothing actually explicitly "disabled" this -- they were not flushed due to a bug. If there is reason to disable cache flushing in HPC-GAP, then the way to do that is to insert code into `FlushCaches`, say inside of `if IsHPCGAP then`, and issue an error or warning message, instead of flushing.

**Comment by fingolfin** [comment]
Looking at the available flush methods, the one for `MemoizePosIntFunction` of course goes to great lengths to be thread safe.  The only two other such methods in the library are installed by each call to `InstallFlushableValue` resp. `InstallFlushableValueFromFunction`. Nothing uses `InstallFlushableValueFromFunction` (except for our test suite). In general, I would try to get rid of both of them, and also `InstallValue`, because they rely on `CLONE_OBJ`. *That* is indeed a serious problem in multi threaded code -- but at the same time, I strongly suspect that most (if not all) code using `InstallFlushableValue` does not actually rely on that behavior. Actually, to the contrary, I think some code could *break* due to this, if `FlushCaches` was called at a bad moment -- so, yeah, perhaps that's a reason to disable cache flushing -- but I think a much better fix wold be to change `InstallFlushableValue` to not use `CLONE_OBJ`, which would solve two problems at once. See also issue #1637, where I just added some remarks on this, too.

**Comment by fingolfin** [comment]
To explain why this change fixes `FlushCaches`, consider this: `FlushCaches` is a simple operation; to make use of it, one does something like this: `InstallMethod( FlushCaches, "return method", [], function() DoSomeFlushing();  TryNextMethod(); end );`. Note that `TryNextMethod` at the end, which is crucial, it chains the separate methods together. The last method being called then is the first  `InstallMethod( FlushCaches, "return method", [], function() end );` which has no `TryNextMethod` call in it.

All methods one install here have the same rank. For methods of the same rank, GAP orders the more recently installed before the previously installed ones. This means that the above `"return method"`, which is the very first method installed for `FlushCaches`, always gets called last.

... except that in HPC-GAP, we never flushed the method caches for `FlushCaches`. So the first time somebody called `FlushCaches`, we recorded some method as being applicable. Then on any subsequent call, we look at the method cache, and end up using that very same method. Never mind that perhaps new methods were installed in the meantime, which rank higher. We keep using the outdated cache.

Since this PR fixes flushing of method caches, it also fixes `FlushCaches`. Here is a much simple test for this. In GAP:
```
gap> DeclareOperation("Foobar", []);
gap> InstallMethod(Foobar, [], {} -> 1);
gap> InstallMethod(Foobar, [], {} -> 2);
gap> Foobar();
2
gap> InstallMethod(Foobar, [], {} -> 3);
gap> Foobar();
3
```
The same in HPC-GAP without this PR:
```
gap> DeclareOperation("Foobar", []);
gap> InstallMethod(Foobar, [], {} -> 1);
gap> InstallMethod(Foobar, [], {} -> 2);
gap> Foobar();
2
gap> InstallMethod(Foobar, [], {} -> 3);
gap> Foobar();
2   # ooops, new method is not used
```

I suspect many more things are silently broken due to this. In fact, I am surprised that there are not some catastrophic breakages caused by this -- probably because for most operations, we install all methods before ever calling them. But I suspect some issues would show up as soon as one uses packages which install new methods for old operations; my guess is that few people us HPC-GAP, and even less with packages that do this, which is why nobody noticed so far?!

File: src/opers.c [hunk 1]
```diff
@@ -1796,6 +1796,9 @@ static inline Obj CacheOper(Obj oper, UInt i)
 
     if (cache == 0) {
         len = (i < 7 ? CACHE_SIZE * (i + 2) : CACHE_SIZE * (1 + 2));
```
Comment (stevelinton) [inline src/opers.c]: What is going on here. I'm finding it hard to see why this line does anything.

File: src/opers.c [hunk 2]
```diff
@@ -1796,6 +1796,9 @@ static inline Obj CacheOper(Obj oper, UInt i)
...
+#ifdef HPCGAP
```
Comment (stevelinton) [inline src/opers.c]: Is it worth the complexity to save one word in non-HPC GAP. Why not just leave this word unused in regular GAP and save on code complexity. There might be a case for reorganising the method cache to  make more efficient use of the CPU cache, but that's a different problem

File: src/opers.c [hunk 3]
```diff
@@ -3540,7 +3547,7 @@ Obj FuncCHANGED_METHODS_OPERATION (
     n = INT_INTOBJ( narg );
     cacheBag = CacheOper( oper, (UInt) n );
     cache = ADDR_OBJ( cacheBag );
-    for ( i = 0;  i < SIZE_OBJ(cacheBag) / sizeof(Obj);  i++ ) {
+    for ( i = 1;  i < SIZE_OBJ(cacheBag) / sizeof(Obj);  i++ ) {
```
Comment (stevelinton) [inline src/opers.c]: This should probably be a memset.

File: src/opers.c (hunk 1 above)
Comment (stevelinton) [reply src/opers.c]: Oh, my bad, I misread the `1` in the else part as another `i`. This line is fine

**Comment by stevelinton** [comment]
I've  merged this. The outstanding suggestions I have are cosmetic and I'd like to see if I can also merge the method reordering PR which is a bit more impactful.

**Pull request merged**